### PR TITLE
Proposal: Add RDS IAM auth

### PR DESCRIPTION
**Pull request by Jell**
### Proposal

natively support IAM authentication towards RDS:

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html

This is the relevant commit: https://github.com/cyrilgdn/terraform-provider-postgresql/commit/315e2d81d64c6196f8ea1e89f23c8dca4f13a09e

This is a proof of concept, tested on our end (as in, used in production :p ) and fully backward compatible. We'd love to contribute into getting this feature merged! For now we've published a fork for now (https://registry.terraform.io/providers/Zimpler/postgresql/latest?pollNotifications=true)

### How to use it:

Given you have an RDS database, with a `root` user that has the `rds_iam` role, you can configure the provider with the following:

```tf
provider "postgresql" {
  scheme   = "awspostgres"
  host     = "my-database.xxxxxxxxxxxx.eu-central-1.rds.amazonaws.com"
  port     = 5432
  username = "root"
  aws_rds_iam_auth = true
  superuser = false
}
```

And this will connect to the database using a token generated from the AWS credentials as per https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/

### Next Step?

Would appreciate feedback on the idea, then if you're OK with it we can refine the code, add documentation, changelog and so on.

Cheers!

### CODE REVIEW

**Comment by bobziuchkovski** [comment]
@Jell Thanks for submitting the PR!  We would also love this support on our end. One additional request, though: would it be possible to add an additional parameter for the AWS profile to use?  Something like `aws_rds_iam_profile`?

Rough sketch, I think it would look something like:

```diff
diff --git a/postgresql/provider.go b/postgresql/provider.go
index ebd1e7f..7e9f9fc 100644
--- a/postgresql/provider.go
+++ b/postgresql/provider.go
@@ -71,6 +71,13 @@ func Provider() terraform.ResourceProvider {
          "(see: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html)",
      },
 
+     "aws_rds_iam_profile": {
+       Type:        schema.TypeString,
+       Optional:    true,
+       Default:     "default",
+       Description: "AWS profile to use for IAM auth",
+     },
+
      // Conection username can be different than database username with user name mapas (e.g.: in Azure)
      // See https://www.postgresql.org/docs/current/auth-username-maps.html
      "database_username": {
@@ -184,7 +191,8 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
  version, _ := semver.ParseTolerant(versionStr)
 
  ctx := context.Background()
- awsConfig, err := awsConfig.LoadDefaultConfig(ctx)
+ profile := d.Get("aws_rds_iam_profile").(string)
+ awsConfig, err := awsConfig.LoadDefaultConfig(ctx, awsConfig.WithSharedConfigProfile(profile))
  if err != nil {
    return nil, err
  }
```

**Comment by Jell** [comment]
Nice, glad it might be useful to others @bobziuchkovski! Note that one could already set the `AWS_PROFILE` environment variable for this, do you think that'd cover your use case? I guess there could be a use case where more than one profile needs to be used in the same terraform state :thinking: in which case maybe worth adding that option

**Comment by cyrilgdn** [comment]
Hi @Jell ,

Thanks a lot for your work on that and sorry for the response delay.

I love this feature and would be happy to merge it in this provider :blush: .
I've just tested it and was able to make it work :+1: (it was more to understand how to configure and RDS instance for that  than to really test the feature as you already use it)

If you want to add @bobziuchkovski patch for AWS profile, it'd be great. It could indeed help in case someone configure multiple  providers in the same state.

Don't forget to merge master, there are some conflicts as you can see, mainly because the `vendor` directory has been removed recently.

**Comment by Jell** [comment]
Also FYI had to apply https://github.com/cyrilgdn/terraform-provider-postgresql/pull/146 to get the linting action to work

**Comment by Jell** [comment]
anything I can do to help move this forward?

File: go.mod [hunk 1]
```diff
@@ -18,7 +20,7 @@ require (
 	github.com/agext/levenshtein v1.2.2 // indirect
 	github.com/apparentlymart/go-textseg v1.0.0 // indirect
 	github.com/apparentlymart/go-textseg/v13 v13.0.0 // indirect
-	github.com/aws/aws-sdk-go v1.36.1 // indirect
+	github.com/aws/aws-sdk-go-v2 v1.9.0
```
Comment (cyrilgdn) [inline go.mod:23]: This should be moved in the first require then

**Review by cyrilgdn — CHANGES_REQUESTED** [review]
@Jell Could you just merge `origin/master` one more time (I tried to do it but maintainers are not allow to update you PR) ? Then we can merge :+1:

File: go.mod (hunk 1 above)
Comment (Jell) [reply go.mod:23]: ah right, will fix

Comment (Jell) [reply go.mod]: I did a rebase and fixed the above, I can do another rebase once https://github.com/cyrilgdn/terraform-provider-postgresql/pull/146 is merged @cyrilgdn and we should be all set!

**Comment by Jell** [comment]
just rebased on top of the ci-lint fix now @cyrilgdn

**Review by cyrilgdn — APPROVED** [review]
@Jell Nice, thanks again for your work :+1:

**Comment by cyrilgdn** [comment]
I'll quickly check other PRs to see if I can include them, but I'll trigger a new release in the next couple of days.

**Comment by bobziuchkovski** [comment]
Thanks for seeing this through, @Jell @cyrilgdn. Much appreciated on my end! ❤️

**Comment by cyrilgdn** [comment]
This has just been released in [v1.15.0](https://registry.terraform.io/providers/cyrilgdn/postgresql/latest)

**Comment by keymon** [comment]
This is great. 

I am missing the capability to assume a role before generating the RDS IAM temporary password. Maybe a `aws_rds_iam_role`.

Actually even 2 roles, because in our use case we have:
 1. CI runs terraform with some AWS IAM user with some credentials
 2. terraform AWS provider assumes a IAM role into the right account.
 3. for RDS, we have other specific IAM role for DB access, which the role in 2 can assume.
 
 So the configuration could be:
 
 ```
 provider "postgresql" {
  scheme   = "awspostgres"
  host     = "my-database.xxxxxxxxxxxx.eu-central-1.rds.amazonaws.com"
  port     = 5432
  username = "root"
  aws_rds_iam_auth = true
  aws_rds_iam_role = "arn:aws:iam::123456789023:role/my_database_master_user"
  aws_assume_role = "arn:aws:iam::123456789023:role/terraform-admin"
  superuser = false
}
```

File: postgresql/provider.go [hunk 2]
```diff
@@ -170,12 +212,28 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
...
 
+	host := d.Get("host").(string)
+	port := d.Get("port").(int)
+	username := d.Get("username").(string)
+
+	var password string
+	if d.Get("aws_rds_iam_auth").(bool) {
+		profile := d.Get("aws_rds_iam_profile").(string)
+		var err error
+		password, err = getRDSAuthToken(profile, username, host, port)
```
Comment (keymon) [inline postgresql/provider.go:223]: Just to note, this password has a limited TTL, I think 15m.

This approach of setting the pwd here should be OK if the terraform execution is shorter.

But ideally we should generate the password every time we create a new DB connection. Maybe here: https://github.com/cyrilgdn/terraform-provider-postgresql/blob/master/postgresql/config.go#L261

To wire this, maybe we could add a callback to the Config object, something like `onConnect` or so. The callback would be called before connect.

Other drivers have hooks for this purpose: https://pkg.go.dev/github.com/jackc/pgx/v4/stdlib#OptionBeforeConnect

**Comment by keymon** [comment]
Other problem with this approach is that you must use the `host` or `PGHOST` of the DB to generate the right credentials. 

But sometimes, to access an RDS instance, one must setup some tunnel, so the connection goes to `localhost`.

t would be great to have an additional flag:
```
aws_rds_iam_host = "..."
```
to override the host with the actual RDS endpoint if it differs from `host`

**Comment by keymon** [comment]
One workaround to the problem [I described above](https://github.com/cyrilgdn/terraform-provider-postgresql/pull/134#issuecomment-1278317064)  is use the profiles of aws cli: 
```
[ profile terraform-admin]
role_arn = arn:aws:iam::123456789023:role/terraform-admin
source_profile = ci-user

[profile my_database_master_user]
role_arn = arn:aws:iam::123456789023:role/my_database_master_user
source_profile = terraform-admin
region = us-west-2
```

Then configure the provider as:
```
provider "postgresql" {
 scheme   = "awspostgres"
 host     = "my-database.xxxxxxxxxxxx.eu-central-1.rds.amazonaws.com"
 port     = 5432
 username = "root"
 # IMPORTANT THIS IS A SUGGESTION. NOT ACTUAL WORKING CONFIG
 aws_rds_iam_auth = true
 aws_rds_iam_profile = "my_database_master_user"
 superuser = false
}

```

that might work, but I was unable to test due the `host` problem

**Pull request merged**